ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+                    for (const auto& common_label : common_labels)
+                    {
+                        common_part += common_label;
+                    }
+                    std::string resultant_subscript = common_part + separate_part1 + separate_part2;
+                    Shape back_shape;
+                    back_shape.insert(
+                        back_shape.end(), common_sub_shape.begin(), common_sub_shape.end());
+                    back_shape.insert(
+                        back_shape.end(), separate1_sub_shape.begin(), separate1_sub_shape.end());
+                    back_shape.insert(
+                        back_shape.end(), separate2_sub_shape.begin(), separate2_sub_shape.end());
+
+                    HostTensorPtr contract_output = std::shared_ptr<HostTensor>(
+                        new HostTensor(matmul_output->get_element_type(), back_shape));
+                    const AxisVector order = get_default_order(matmul_output->get_shape());
+                    ngraph::runtime::reference::reshape(
+                        reinterpret_cast<const char*>(matmul_output->get_data_ptr<T>()),
+                        reinterpret_cast<char*>(contract_output->get_data_ptr<T>()),
+                        matmul_output->get_shape(),
+                        order,
+                        back_shape,
+                        matmul_output->get_element_type().size());
+
+                    update_operands(inputs,
+                                    input_subscripts,
+                                    input_ind1,
+                                    input_ind2,
+                                    contract_output,
+                                    resultant_subscript);
+                }
+
+                template <typename T>
+                void einsum_impl(const HostTensorVector& inputs,
+                                 const HostTensorVector& outputs,
+                                 const std::string& equation)
+                {
+                    std::vector<std::string> input_subscripts;
+                    std::string output_subscript;
+                    ngraph::opset7::Einsum::parse_equation(
+                        equation, input_subscripts, output_subscript);
+
+                    // compute einsum path that is used to contract a pair of operands
+                    // in more optimal order
+                    size_t num_inputs = inputs.size();
+                    auto einsum_path = compute_einsum_path(num_inputs);
+
+                    HostTensorVector int_inputs = inputs;
+
+                    // contract inputs by Einsum until just one is remained
+                    for (auto const& inds_pair : einsum_path)
+                    {
+                        contract_two_inputs<T>(int_inputs,
+                                               input_subscripts,
+                                               output_subscript,
+                                               inds_pair.first,
```

Comment:
As I understand `inds_pair.first` is always equal to `0`, do we need to have this pair?

Reply:
Yes, we do because `compute_einsum_path` is now undone. And it will be complete once we receive model request with several operand Einsum.